File: xml-inflater/src/main/java/com/itsaky/inflater/adapters/android/widget/GridLayoutAttrAdapter.java
/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.inflater.adapters.android.widget;

import android.util.DisplayMetrics;
import android.view.View;
import android.widget.GridLayout;

import androidx.annotation.NonNull;

import com.itsaky.inflater.IAttribute;
import com.itsaky.inflater.IResourceTable;
import com.itsaky.inflater.adapters.android.view.ViewGroupAttrAdapter;

/**
 * @author Akash Yadav
 */
public class GridLayoutAttrAdapter extends ViewGroupAttrAdapter {

  public GridLayoutAttrAdapter(
      @NonNull IResourceTable resourceFinder, DisplayMetrics displayMetrics) {
    super(resourceFinder, displayMetrics);
  }

  @Override
  public boolean isApplicableTo(View view) {
    return view instanceof GridLayout;
  }

  @Override
  public boolean apply(@NonNull IAttribute attribute, @NonNull View view) {
    final var grid = (GridLayout) view;
    final var name = attribute.getAttributeName();
    final var value = attribute.getValue();

    if (!canHandleNamespace(attribute)) {
      return false;
    }

    boolean handled = true;
    switch (name) {
      case "alignmentMode":
        grid.setAlignmentMode(parseAlignmentMode(value));
        break;
      case "columnCount":
        grid.setColumnCount(parseInteger(value, Integer.MIN_VALUE));
        break;
      case "columnOrderPreserved":
        grid.setColumnOrderPreserved(parseBoolean(value));
        break;
      case "orientation":
        grid.setOrientation(parseOrientation(value));
        break;
      case "rowCount":
        grid.setRowCount(parseInteger(value, Integer.MIN_VALUE));
        break;
      case "rowOrderPreserved":
        grid.setRowOrderPreserved(parseBoolean(value));
        break;
      case "useDefaultMargins":
        grid.setUseDefaultMargins(parseBoolean(value));
        break;
      default:
        handled = false;
        break;
    }

    if (!handled) {
      handled = super.apply(attribute, view);
    }

    return handled;
  }

  protected int parseOrientation(String value) {
    if ("vertical".equals(value)) {
      return GridLayout.VERTICAL;
    }

    return GridLayout.HORIZONTAL;
  }

  protected int parseAlignmentMode(String value) {
    if ("alignBounds".equals(value)) {
      return GridLayout.ALIGN_BOUNDS;
    }

    return GridLayout.ALIGN_MARGINS;
  }
}

File: xml-inflater/src/main/java/com/itsaky/androidide/inflater/internal/adapters/GridLayoutAttrAdapter.kt
/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.androidide.inflater.internal.adapters

import android.widget.GridLayout
import com.itsaky.androidide.annotations.inflater.ViewAdapter
import com.itsaky.androidide.inflater.AttributeHandlerScope

/**
 * Attribute adapter for [GridLayout].
 *
 * @author Akash Yadav
 */
@ViewAdapter(GridLayout::class)
open class GridLayoutAttrAdapter<T : GridLayout> : ViewGroupAttrAdapter<T>() {
  override fun createAttrHandlers(create: (String, AttributeHandlerScope<T>.() -> Unit) -> Unit) {
    super.createAttrHandlers(create)
    create("alignmentMode") { view.alignmentMode = parseAlignmentMode(value) }
    create("columnCount") { view.columnCount = parseInteger(value, Int.MIN_VALUE) }
    create("columnOrderPreserved") { view.isColumnOrderPreserved = parseBoolean(value) }
    create("orientation") { view.orientation = parseOrientation(value) }
    create("rowCount") { view.rowCount = parseInteger(value, Int.MIN_VALUE) }
    create("rowOrderPreserved") { view.isRowOrderPreserved = parseBoolean(value) }
    create("useDefaultMargins") { view.useDefaultMargins = parseBoolean(value) }
  }

  protected open fun parseOrientation(value: String): Int {
    return if ("vertical" == value) {
      GridLayout.VERTICAL
    } else GridLayout.HORIZONTAL
  }

  protected open fun parseAlignmentMode(value: String): Int {
    return if ("alignBounds" == value) {
      GridLayout.ALIGN_BOUNDS
    } else GridLayout.ALIGN_MARGINS
  }
}
